<?xml version="1.0"?>
<robot name="armatronix">

    <link name="base_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <box size="0.7 0.6 0.15" />
            </geometry>
            <material name="">
                <color rgba="0.79216 0.81961 0.93333 1" />
            </material>
        </visual>
    <collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><box size="0.7 0.6 0.15" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="5" /><inertia ixx="0.159375" ixy="0.0" ixz="0.0" iyy="0.213542" iyz="0.0" izz="0.354167" /></inertial><collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><box size="0.7 0.6 0.15" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="5" /><inertia ixx="0.159375" ixy="0.0" ixz="0.0" iyy="0.213542" iyz="0.0" izz="0.354167" /></inertial></link>


    <link name="camera_link">

    </link>

    <joint name="camera_joint" type="fixed">
        <axis xyz="0 1 0" />
        <parent link="base_link" />
        <child link="camera_link" />
        <origin rpy="0 0 0" xyz="0.275 0 0.017" />
    </joint>


    <link name="front_left_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="0 1.5708 0" />
            <geometry>
                
                <cylinder radius="0.1" length="0.05" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1.0" />
            </material>
        </visual>
    <collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><cylinder radius="0.1" length="0.05" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="2" /><inertia ixx="0.005417" ixy="0.0" ixz="0.0" iyy="0.005417" iyz="0.0" izz="0.010000" /></inertial><collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><cylinder radius="0.1" length="0.05" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="2" /><inertia ixx="0.005417" ixy="0.0" ixz="0.0" iyy="0.005417" iyz="0.0" izz="0.010000" /></inertial></link>

    <joint name="front_left_wheel_joint" type="continuous">
        <axis xyz="1 0 0" />
        <parent link="base_link" />
        <child link="front_left_wheel" />
        <origin xyz="0.3 0.325 0" rpy="0 0 1.57" />
    </joint>


    <link name="front_right_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="0 4.7129 0" />
            <geometry>
                <cylinder radius="0.1" length="0.05" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1.0" />
            </material>
        </visual>
    <collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><cylinder radius="0.1" length="0.05" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="2" /><inertia ixx="0.005417" ixy="0.0" ixz="0.0" iyy="0.005417" iyz="0.0" izz="0.010000" /></inertial></link>

    <joint name="front_right_wheel_joint" type="continuous">
        <axis xyz="1 0 0" />
        <parent link="base_link" />
        <child link="front_right_wheel" />
        <origin xyz="0.3 -0.325 0" rpy="0 0 1.57" />
    </joint>


    <link name="rear_left_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="0 1.5708 0" />
            <geometry>
                <cylinder radius="0.1" length="0.05" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1.0" />
            </material>
        </visual>
    <collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><cylinder radius="0.1" length="0.05" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="2" /><inertia ixx="0.005417" ixy="0.0" ixz="0.0" iyy="0.005417" iyz="0.0" izz="0.010000" /></inertial></link>

    <joint name="rear_left_wheel_joint" type="continuous">
        <axis xyz="1 0 0" />
        <parent link="base_link" />
        <child link="rear_left_wheel" />
        <origin xyz="-0.3 0.325 0" rpy="0 0 1.57" />
    </joint>


    <link name="rear_right_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="3.14 1.5708 0" />
            <geometry>
                <cylinder radius="0.1" length="0.05" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1.0" />
            </material>
        </visual>
    <collision><origin xyz="0 0 0" rpy="0 0 0" /><geometry><cylinder radius="0.1" length="0.05" /></geometry></collision><inertial><origin xyz="0 0 0" rpy="0 0 0" /><mass value="2" /><inertia ixx="0.005417" ixy="0.0" ixz="0.0" iyy="0.005417" iyz="0.0" izz="0.010000" /></inertial></link>

    <joint name="rear_right_wheel_joint" type="continuous">
        <axis xyz="1 0 0" />
        <parent link="base_link" />
        <child link="rear_right_wheel" />
        <origin xyz="-0.3 -0.325 0" rpy="0 0 1.57" />
    </joint>





    <link name="arm_base_link">
    <visual>
        <geometry>
            <cylinder length="0.05" radius="0.2" />
        </geometry>
        <material name="Black">
            <color rgba="0 0 0 1" />
        </material>
        <origin rpy="0 0 0" xyz="0 0 0.025" />
    </visual>
    <collision>
        <geometry>
            <cylinder length="0.05" radius="0.2" />
        </geometry>
        <origin xyz="0 0 0.025" rpy="0 0 0" />
    </collision>
    <inertial>
        <mass value="2.0" /> <!-- Adjusted mass considering the base -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.0035" ixy="0.0" ixz="0.0" iyy="0.0035" iyz="0.0" izz="0.007" />
    </inertial>
</link>

<link name="link_1">
    <visual>
        <origin xyz="0.0 0.0 0.25" rpy="0.0 0.0 0.0" />
        <geometry>
            <cylinder radius="0.08" length="0.5" />
        </geometry>
        <material name="Blue">
            <color rgba="0 0 0.8 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.08" length="0.5" />
        </geometry>
        <origin xyz="0.0 0.0 0.25" rpy="0.0 0.0 0.0" />
    </collision>
    <inertial>
        <mass value="1.5" /> <!-- Adjusted for the long link -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.005" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.0015" />
    </inertial>
</link>

<link name="link_2">
    <visual>
        <geometry>
            <cylinder radius="0.08" length="0.1" />
        </geometry>
        <material name="Red">
            <color rgba="1 0 0 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.08" length="0.1" />
        </geometry>
        <origin xyz="0 0 0.05" rpy="0 0 0" />
    </collision>
    <inertial>
        <mass value="0.5" /> <!-- Lighter link due to its size -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.0003" ixy="0.0" ixz="0.0" iyy="0.0003" iyz="0.0" izz="0.0001" />
    </inertial>
</link>

<link name="link_3">
    <visual>
        <geometry>
            <cylinder radius="0.05" length="0.4" />
        </geometry>
        <material name="Blue">
            <color rgba="0.5 0.5 0.5 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.05" length="0.4" />
        </geometry>
        <origin xyz="0 0 0.2" rpy="0 0 0" />
    </collision>
    <inertial>
        <mass value="1.0" /> <!-- Moderate mass for stability -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.0015" ixy="0.0" ixz="0.0" iyy="0.0015" iyz="0.0" izz="0.0006" />
    </inertial>
</link>

<link name="link_4">
    <visual>
        <geometry>
            <cylinder radius="0.06" length="0.1" />
        </geometry>
        <material name="Red">
            <color rgba="1 0 0 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.06" length="0.1" />
        </geometry>
        <origin xyz="0 0 0.05" rpy="0 0 0" />
    </collision>
    <inertial>
        <mass value="0.5" /> <!-- Small link, lower mass -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.0002" ixy="0.0" ixz="0.0" iyy="0.0002" iyz="0.0" izz="0.00005" />
    </inertial>
</link>

<link name="link_5">
    <visual>
        <geometry>
            <cylinder radius="0.04" length="0.3" />
        </geometry>
        <material name="Blue">
            <color rgba="0.5 0.5 0.5 1" />
        </material>
    </visual>
    <collision>
        <geometry>
            <cylinder radius="0.04" length="0.3" />
        </geometry>
        <origin xyz="0 0 0.15" rpy="0 0 0" />
    </collision>
    <inertial>
        <mass value="0.8" /> <!-- Final link with payload capability -->
        <origin xyz="0 0 0" rpy="0 0 0" />
        <inertia ixx="0.0009" ixy="0.0" ixz="0.0" iyy="0.0009" iyz="0.0" izz="0.0002" />
    </inertial>
</link>

<link name="gripper_link">
    <visual>
        <geometry>
            <mesh filename="package://armatronix/meshes/angular_gripper.stl" scale="0.002 0.002 0.002"/>

        </geometry>
        <material name="black">
                <color rgba="0 0 0 1.0" />
        </material>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </visual>

    <collision>
        <geometry>
            <mesh filename="package://armatronix/meshes/angular_gripper.stl" scale="0.002 0.002 0.002"/>

        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </collision>

    <inertial>
        <mass value="0.3"/>  <!-- Approximate mass of the gripper, you can adjust this -->
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
</link>


<joint name="joint_0" type="fixed">
    <parent link="base_link" />
    <child link="arm_base_link" />
    <origin rpy="0 0 0" xyz="0.0 0.0 0.08" />
</joint>

<joint name="joint_1" type="continuous">
    <axis xyz="0 0 1" />
    <parent link="arm_base_link" />
    <child link="link_1" />
    <origin rpy="0 0 0" xyz="0.0 0.0 0.05" />
</joint>

<joint name="joint_2" type="revolute">
    <axis xyz="0 0 1" />
    <parent link="link_1" />
    <child link="link_2" />
    <origin rpy="0 1.5708 0" xyz="0.0 -0.005 0.58" />
    <limit lower="-0.127" upper="1.57" effort="10" velocity="0.5" />
</joint>

<joint name="joint_3" type="fixed">
    <parent link="link_2" />
    <child link="link_3" />
    <origin rpy="1.57 0 0" xyz="0.0 0.2 0 " />
</joint>

<joint name="joint_4" type="revolute">
    <parent link="link_3" />
    <child link="link_4" />
    <origin rpy="1.57 0 0" xyz=" 0 0 -0.25" />
    <axis xyz=" 0 0 1" />
    <limit lower="-0.00" upper="1.57" effort="10" velocity="0.5" />
</joint>

<joint name="joint_5" type="fixed">
    <parent link="link_4" />
    <child link="link_5" />
    <origin rpy="1.57 0 0" xyz="0.0 -0.2 0 " />
</joint>

<joint name="gripper_joint" type="continuous">
    <axis xyz="0 1 0" />
    <parent link="link_5"/>
    <child link="gripper_link"/>
    <origin xyz="0 0 0.21" rpy="1.5708 0 0"/>  <!-- Adjust based on where the gripper should be attached -->
    
</joint>


<gazebo reference="arm_base_link">
    <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="link_1">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="link_3">
    <material>Gazebo/Orange</material>
</gazebo>

<gazebo reference="link_2">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="link_4">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="link_5">
    <material>Gazebo/Orange</material>
</gazebo>

</robot>





    <!-- <?xml version="1.0"?> -->
<!-- <robot name="armatronix"> -->
    <!-- <link name="base_link">
        <visual>
            <geometry>
                <cylinder length="0.05" radius="0.2"/>
            </geometry>
            <material name="Black">
                <color rgba="0 0 0 1"/>
            </material>
            <origin rpy="0 0 0" xyz="0 0 0.025"/>
        </visual>
        
        <collision>
            <geometry>
                <cylinder radius="0.2" length="0.05"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0.025"/>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.025" rpy="0.0 0.0 0.0"/>
            <mass value="5.0"/>
            <inertia ixx="0.0135" ixy="0.0" ixz="0.0" iyy="0.0135" iyz="0.0" izz="0.05"/>
        </inertial>
    </link>

    <link name="link_1">
        <visual>
            <origin xyz="0.0 0.0 0.25" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.08" length="0.5"/>
            </geometry>
            <material name="Blue">
                <color rgba="0 0 0.8 1"/>
            </material>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 0.25" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.08" length="0.5"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.25" rpy="0.0 0.0 0.0"/>
            <mass value="5.0"/>
            <inertia ixx="0.107" ixy="0.0" ixz="0.0" iyy="0.107" iyz="0.0" izz="0.0125"/>
        </inertial>
    </link>
        

    <link name="link_2">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.1"/>
            </geometry>
            <material name="Red">
                <color rgba="1 0 0 1"/>
            </material>
        </visual>

        <collision>
            <geometry>
                <cylinder radius="0.08" length="0.1"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.2" rpy="0.0 0.0 0.0"/>
            <mass value="2.0"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>
        

    <link name="link_3">
        <visual>
            <geometry>
                <cylinder radius="0.05" length="0.4"/>
            </geometry>
            <material name="Blue">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>

        <collision>
            <geometry>
                <cylinder radius="0.05" length="0.4"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.2" rpy="0.0 0.0 0.0"/>
            <mass value="0.01"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>


    <link name="link_4">
        <visual>
            <geometry>
                <cylinder radius="0.06" length="0.1"/>
            </geometry>
            <material name="Red">
                <color rgba="1 0 0 1"/>
            </material>
        </visual>

        <collision>
            <geometry>
                <cylinder radius="0.06" length="0.1"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.2" rpy="0.0 0.0 0.0"/>
            <mass value="0.01"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>


    <link name="link_5">
        <visual>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <material name="Yelow">
                <color rgba="0 1 0.5 1"/>
            </material>
        </visual>

        <collision>
            <geometry>
                <cylinder radius="0.03" length="0.3"/>
            </geometry>
            <dynamics damping="0.0" friction="0.0"/>
        </collision>

        <inertial>
            <origin xyz="0.0 0.0 0.2" rpy="0.0 0.0 0.0"/>
            <mass value="0.01"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>



    <joint name="joint_1" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="base_link"/>
        <child link="link_1"/>
        <origin rpy="0 0 0" xyz="0.0 0.0 0.05"/>

    </joint>

    <joint name="joint_2" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="link_1"/>
        <child link="link_2"/>
        <origin rpy="0 1.5708 0" xyz="0.0 -0.005 0.58"/>
        <limit lower="-0.25" upper="3.34" effort="10" velocity="0.5"/>

    </joint>

    <joint name="joint_3" type="fixed">
        <parent link="link_2"/>
        <child link="link_3"/>
        <origin rpy="1.57 0 0" xyz="0.0 0.2 0 "/>

    </joint>


    <joint name="joint_4" type="continuous">
        <parent link="link_3"/>
        <child link="link_4"/>
        <origin rpy="1.57 0 0" xyz=" 0 0 -0.25"/>
        <axis xyz=" 0 0 1"/>
        <limit lower="-1.92" upper="1.92" effort="10" velocity="0.5"/>

    </joint>

    <joint name="joint_5" type="fixed">
        <parent link="link_4"/>
        <child link="link_5"/>
        <origin rpy="1.57 0 0" xyz="0.0 -0.2 0 "/>

    </joint>


    <gazebo reference="base_link">
    <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="link_1">
    <material>Gazebo/Orange</material>
    </gazebo>

    <gazebo reference="link_3">
    <material>Gazebo/Orange</material>
    </gazebo>

    <gazebo reference="link_2">
    <material>Gazebo/Red</material>
    </gazebo>

    <gazebo reference="link_4">
    <material>Gazebo/Red</material>
    </gazebo>

    <gazebo reference="link_5">
    <material>Gazebo/Orange</material>
    </gazebo>


    <ros2_control name="GazeboSystem" type="system">
    <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="joint_1">

        <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
        </command_interface>

        <state_interface name="position">
        <param name="initial_value">1.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>

    </joint>
    </ros2_control>

    <ros2_control name="GazeboSystem" type="system">
    <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="joint_2">

        <command_interface name="position">
        <param name="min">-0.75</param>
        <param name="max">3.14</param>
        </command_interface>

        <state_interface name="position">
        <param name="initial_value">1.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>

    </joint>
    </ros2_control>

    <ros2_control name="GazeboSystem" type="system">
    <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="joint_4">

        <command_interface name="position">
        <param name="min">-2.00</param>
        <param name="max">1.57</param>
        </command_interface>

        <state_interface name="position">
        <param name="initial_value">1.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>

    </joint>
    </ros2_control>


    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
        <robot_param>robot_description</robot_param>
        <robot_param_node>robot_state_publisher</robot_param_node>
        <parameters>/home/luqman/udemy_ws/src/robotic_arms_control/config/bazu_controller.yaml</parameters>
        </plugin>
    </gazebo>

</robot> -->

